src/visitable.cpp

```diff
@@ -72,21 +72,30 @@ bool visitable<T>::has_item_with( const std::function<bool( const item & )> &fil
     } ) == VisitResponse::ABORT;
 }
 
+// Sums the two terms, being careful to not trigger any overflows.
+// Assumes that all terms are positive and base <= limit.
+static long accumulate_with_limit( long base, long addend, long limit ) {
+    if( addend > limit - base ) {
+        return limit;
+    }
+    return base + addend;
+}
+
 template <typename T>
 static int has_quality_internal( const T& self, const quality_id &qual, int level, int limit )
 {
     int qty = 0;
 
     self.visit_items( [&qual, level, &limit, &qty]( const item *e ) {
         if( e->get_quality( qual ) >= level ) {
-            qty += e->count_by_charges() ? e->charges : 1;
+            qty = accumulate_with_limit( qty, e->count_by_charges() ? e->charges : 1, limit );
```

Comment:
Whilst the `visitable` class reduced duplication it's still complex and the problem is this fix is fairly obfuscating. For our purposes an `int` is sufficiently large - the overflow only really occurs when we deliberately add `INT_MAX` for infinite charges. A conceptually simpler fix would just be to add a million charges or something similar - it wouldn't be perfect but it would fix all the current bugs without obfuscating.